### PR TITLE
added react-alt example

### PR DESCRIPTION
**Pull request by wishpishh**
I've created an example using React and the Alt implementation of the Flux pattern. It's based on the React example already provided by Pete Hunt, but modified to use Alt instead of a custom model. There is currently not a Flux example, and Alt is one of the currently most popular and widely used implementations.

### CODE REVIEW

**Comment by addyosmani** [comment]
Thanks for your PR. I'm familiar with React and Flux (and Redux etc) but haven't come across Alt before. How does Alt differ from the other Flux pattern implementations? Some more information that would help us figure out if this would be useful to folks would go a long way :)

PS: CI tests are currently failing.

**Comment by wishpishh** [comment]
Seeing as there is no current Flux example at all, I felt that Alt was the best implementation to submit as it is currently one of the more active Flux implementations (https://github.com/goatslacker/alt/pulse vs https://github.com/yahoo/fluxible/pulse vs https://github.com/reflux/refluxjs/pulse vs https://github.com/BinaryMuse/fluxxor/pulse). Redux is great and interesting, but it departs rather sharply from the reference Flux implementation presented by Facebook and is tightly coupled with the React Hot Reloader and Webpack, which is great but requires a deeper look into the dev environment. That makes Alt a more direct and representative example of Flux that lends itself well as an TodoMVC example, in my opinion. Plus, I had to choose one, after all :)

I'm looking into why the CI tests are failing, they're passing when I do it locally and I haven't had more time to investigate how to reproduce the failing test results yet. This is what I quickly did, and everything passed:

cd todomvc/tests
npm install
npm run serve
npm run test

**Comment by samccone** [comment]
No memory leaks detected :+1:

**Comment by samccone** [comment]
@wishpishh mind squashing this down to a single commit for us?

**Comment by wishpishh** [comment]
There you go.

**Comment by passy** [comment]
Just wanted to say that I've seen react-alt a couple times internally and kinda wished there was a TodoMVC example. :)

**Comment by addyosmani** [comment]
@passy has more experience with React and related libraries than I do. Deferring to him for a call on whether this should land.

File: examples/react-alt/js/app.jsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	var TodoApp = React.createClass({
+		getInitialState: function () {
+			return TodoStore.getState();
+		},
+
+		componentDidMount: function () {
+			TodoStore.listen(this.onStoreChange);
+
+			var router = Router({
+				'/': function () {TodoActions.show(app.ALL_TODOS);},
```
Comment (passy) [inline examples/react-alt/js/app.jsx]: Could you add a bit of whitespace here? `function () { TodoActions.show(app.ALL_TODOS); }`. Alternatively, binding the function is fine, too: `TodoActions.show.bind(TodoActions, app.ALL_TODOS);`

File: examples/react-alt/package.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "private": true,
+  "dependencies": {
+    "alt": "^0.17.3",
+    "director": "^1.2.0",
+    "react": "^0.13.3",
```
Comment (passy) [inline examples/react-alt/package.json]: Does this already work with `0.14`?

File: examples/react-alt/readme.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# React + Alt TodoMVC Example
+
+> React is a JavaScript library for creating user interfaces. Its core principles are declarative code, efficiency, and flexibility. Simply specify what your component looks like and React will keep it up-to-date when the underlying data changes. Alt is one of many implementations of Facebook's Flux architecture.
```
Comment (passy) [inline examples/react-alt/readme.md]: Could you point out some attribute that differentiates Alt from the other implementations?

**Comment by samccone** [comment]
@goatslacker || @jdlehman would you mind taking a look at this to make sure it is best using alt?

File: examples/react-alt/js/actions/todoActions.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+/*jshint newcap:false */
+
+var app = app || {};
+
+(function () {
+	'use strict';
+
+	var Utils = app.Utils;
+
+	app.todoActions = app.alt.createActions({
```
Comment (jdlehman) [inline examples/react-alt/js/actions/todoActions.js]: most of this boilerplate could be reduced with [`generateActions`](https://github.com/goatslacker/alt/blob/master/docs/createActions.md#generateactions) as most of your actions are just passing the data through directly. The exception to this is `addTodo`, though I would probably create the default properties for the todo like `id` and `completed` in the store and just have `title` passed through via `generateActions`

something like:
```js
app.todoActions = app.alt.generateActions(
  'addTodo',
  'toggleAll',
  'toggle',
  'destroy',
  'save',
  'clearCompleted',
  'edit',
  'show'
);
```

for this change to work I believe you would also need to pass an object of data in the instances where you have more than one argument.

File: examples/react-alt/js/app.jsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	var TodoFooter = app.TodoFooter;
+	var TodoItem = app.TodoItem;
+	var TodoActions = app.todoActions;
+	var TodoStore = app.todoStore;
+
+	var ENTER_KEY = 13;
+
+	var TodoApp = React.createClass({
+		getInitialState: function () {
+			return TodoStore.getState();
```
Comment (jdlehman) [inline examples/react-alt/js/app.jsx]: you could also use `[connectToStores]`(https://github.com/altjs/connect-to-stores) to wire up store data using a HOC to pass props down to the `App` component rather than directly accessing the global `TodoStore`.

This will also remove the need to listen and unlisten in `componentDidMount` and `componentDidUnmount` and is the recommended way to connect stores to your components.

**Comment by jdlehman** [comment]
I looked over this and made a few comments, but @goatslacker might also have comments.

Also [this](https://github.com/goatslacker/alt/tree/master/examples/todomvc) is the TodoMVC from the examples in alt's repo as a reference (but it is a bit out of date and could use some TLC...)

**Comment by samccone** [comment]
:bow: @jdlehman .. thanks for the review!

File: examples/react-alt/js/actions/todoActions.js (hunk 4 above)
Comment (goatslacker) [inline examples/react-alt/js/actions/todoActions.js]: Please use return to dispatch.

Example

```js
addTodo: function (title) {
  return {
    id: Utils.uuid(),
    title: title,
    completed: false
  }
}
```

File: examples/react-alt/js/app.jsx (hunk 5 above)
Comment (goatslacker) [inline examples/react-alt/js/app.jsx]: Yes please :)

File: examples/react-alt/js/stores/todoStore.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+var app = app || {};
+
+(function () {
+	'use strict';
+
+	var Utils = app.Utils;
+	var LOCALSTORAGE_NAMESPACE = 'react-alt-todo';
+
+	var TodoStore = function () {
+		this.todos = Utils.store(LOCALSTORAGE_NAMESPACE + '.todos');
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]:
```js
this.state = {
  todos: ...,
  nowShowing: ...,
  editing: ...,
};
```

File: examples/react-alt/js/stores/todoStore.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+			destroy: app.todoActions.destroy,
+			save: app.todoActions.save,
+			clearCompleted: app.todoActions.clearCompleted,
+			edit: app.todoActions.edit,
+			show: app.todoActions.show
+		});
+	};
+
+	TodoStore.prototype.addTodo = function (todo) {
+		this.todos = this.todos.concat(todo);
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]: `this.setState({ todos: this.state.todos.concat(todo) })`

File: examples/react-alt/js/stores/todoStore.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+		Utils.store(LOCALSTORAGE_NAMESPACE + '.todos', this.todos);
+	};
+
+	TodoStore.prototype.toggleAll = function (checked) {
+		this.todos = this.todos.map(function (todo) {
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]: setState

File: examples/react-alt/js/stores/todoStore.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+			return Utils.extend({}, todo, {completed: checked});
+		});
+
+		Utils.store(LOCALSTORAGE_NAMESPACE + '.todos', this.todos);
+	};
+
+	TodoStore.prototype.toggle = function (todoToToggle) {
+		this.todos = this.todos.map(function (todo) {
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]: setState

File: examples/react-alt/js/stores/todoStore.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+			return todo !== todoToToggle ?
+				todo :
+				Utils.extend({}, todo, {completed: !todo.completed});
+		});
+
+		Utils.store(LOCALSTORAGE_NAMESPACE + '.todos', this.todos);
+	};
+
+	TodoStore.prototype.destroy = function (todoToDestroy) {
+		this.todos = this.todos.filter(function (todo) {
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]: setState

File: examples/react-alt/js/stores/todoStore.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+			return todo !== todoToDestroy;
+		});
+
+		Utils.store(LOCALSTORAGE_NAMESPACE + '.todos', this.todos);
+	};
+
+	TodoStore.prototype.save = function (command) {
+		this.todos = this.todos.map(function (todo) {
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]: setState

File: examples/react-alt/js/stores/todoStore.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+			return todo !== command.todoToSave ?
+				todo :
+				Utils.extend({}, command.todoToSave, {title: command.text});
+		});
+
+		Utils.store(LOCALSTORAGE_NAMESPACE + '.todos', this.todos);
+	};
+
+	TodoStore.prototype.clearCompleted = function () {
+		this.todos = this.todos.filter(function (todo) {
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]: setState

File: examples/react-alt/js/stores/todoStore.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+			return !todo.completed;
+		});
+
+		Utils.store(LOCALSTORAGE_NAMESPACE + '.todos', this.todos);
+	};
+
+	TodoStore.prototype.edit = function (id) {
+		this.editing = id;
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]: setState

File: examples/react-alt/js/stores/todoStore.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+		Utils.store(LOCALSTORAGE_NAMESPACE + '.editing', this.editing);
+	};
+
+	TodoStore.prototype.show = function (nowShowing) {
+		this.nowShowing = nowShowing;
```
Comment (goatslacker) [inline examples/react-alt/js/stores/todoStore.js]: setState

File: examples/react-alt/package.json (hunk 2 above)
Comment (goatslacker) [inline examples/react-alt/package.json]: should work with 0.14

File: examples/react-alt/readme.md (hunk 3 above)
Comment (goatslacker) [inline examples/react-alt/readme.md]: There's some stuff here: http://alt.js.org/

You can also check this out: https://github.com/goatslacker/alt#why-you-should-be-using-alt

**Comment by goatslacker** [comment]
This is really cool! :sunglasses: looking awesome

File: examples/react-alt/js/app.jsx (hunk 5 above)
Comment (wishpishh) [inline examples/react-alt/js/app.jsx]: My thinking is that this perhaps obscures the example a bit by making it depart more from the current React example. One of my principles when writing this was that it should be as simple as possible to compare to the current React example.

File: examples/react-alt/package.json (hunk 2 above)
Comment (wishpishh) [inline examples/react-alt/package.json]: 0.14 doesn't ship with JSXTransformer, so that's why I didn't upgrade. I tried using Babel's in-browser transformer but didn't get it to immediately work so I gave it up for now.

**Comment by samccone** [comment]
@wishpishh mind squashing this up into one?

File: examples/react-alt/js/app.jsx [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+				return;
+			}
+
+			event.preventDefault();
+
+			var val = React.findDOMNode(this.refs.newField).value.trim();
+
+			if (val) {
+				TodoActions.addTodo(val);
+				React.findDOMNode(this.refs.newField).value = '';
```
Comment (passy) [inline examples/react-alt/js/app.jsx]: Why is this necessary? Why doesn't React take care of this when the state is changed?

File: examples/react-alt/js/app.jsx [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+		cancel: function () {
+			TodoActions.edit(null);
+		},
+
+		clearCompleted: function () {
+			TodoActions.clearCompleted();
+		},
+
+		render: function () {
+			var footer;
```
Comment (passy) [inline examples/react-alt/js/app.jsx]: Could you initialize these instead of implicitly relying on them being `undefined`?

File: examples/react-alt/.gitignore [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+!node_modules/alt/dist
+node_modules/alt/dist/*
+!node_modules/alt/dist/alt.js
+
+node_modules/todomvc-app-css/*
+!node_modules/todomvc-app-css/index.css
+
+node_modules/todomvc-common/*
+!node_modules/todomvc-common/base.css
+!node_modules/todomvc-common/base.js
```
Comment (arthurvr) [inline examples/react-alt/.gitignore]: Nit: missing newline.

File: examples/react-alt/js/actions/todoActions.js [hunk 18]
```diff
@@ -0,0 +1,35 @@
+/*jshint quotmark:false */
+/*jshint white:false */
+/*jshint trailing:false */
```
Comment (arthurvr) [inline examples/react-alt/js/actions/todoActions.js]: `trailing` and `white` are [deprecated](https://github.com/jshint/jshint/releases/tag/2.5.0).

File: examples/react-alt/index.html [hunk 19]
```diff
@@ -0,0 +1,34 @@
+<!doctype html>
+<html lang="en" data-framework="react">
+	<head>
+		<meta charset="utf-8">
+		<title>React + Alt• TodoMVC</title>
```
Comment (arthurvr) [inline examples/react-alt/index.html]: spacing nit: `<title>React + Alt • TodoMVC</title>`

**Comment by arthurvr** [comment]
Wow! This is lovely work.

File: examples/react-alt/js/app.jsx (hunk 15 above)
Comment (wishpishh) [inline examples/react-alt/js/app.jsx]: If I understand you correctly, you're asking why it's necessary to clear the value in the text input?

I could have set the value of the input explicitly from some property in state, but that would turn the component into a so-called [controlled component](http://facebook.github.io/react/docs/forms.html#controlled-components) which would require me to write code to update the state on input change. This is less boilerplate, I guess, although it's up for debate which is clearer. This is also the way it's done in the current React example which this is based on.

Comment (passy) [inline examples/react-alt/js/app.jsx]: Thanks for elaborating. To be honest, I completely missed it whenever the change from a controlled to an uncontrolled component was made in our React example. I think in our case, where we *want* to control the input in a bidirectional way, this is a bad practice. But this needs to be addressed in the base example first and I don't mind you sticking close to the base example.

Comment (wishpishh) [inline examples/react-alt/js/app.jsx]: I'll be happy to update this example to reflect your base React example if it changes, to keep the two easily comparable.

Comment (passy) [inline examples/react-alt/js/app.jsx]: Brilliant!

**Comment by samccone** [comment]
@passy any finalish thoughts :)

**Comment by passy** [comment]
Will do another scan tonight. 👌

File: examples/react-alt/js/app.jsx (hunk 15 above)
Comment (passy) [inline examples/react-alt/js/app.jsx]: I've updated the main React example for this. Would you mind adjusting the PR accordingly?

**Comment by passy** [comment]
The comment I left doesn't seem to show up in the threaded view here. Looks all good, but it would be awesome if you could update the `findDOMNode` to reflect the upstream changes. :)

**Comment by passy** [comment]
@wishpishh Friendly ping. :pager:

**Comment by wishpishh** [comment]
Done :)

**Comment by passy** [comment]
Landed!



Thanks so much for the submission!

**Comment by samccone** [comment]
AWWWWwww YES 

A++++++ work @wishpishh

**Comment by goatslacker** [comment]
:fireworks: :tada: :confetti_ball:
